Sort by proximity: shuffle equal-distance replicas

bhalevy opened the pull request:
This series re-implements locator::topology::sort_by_proximity
and adds some randomization to shuffle equal-distance replicas for improving load-balancing
when reading with 1 < consistency level < replication factor.

This change also adds a manual test for benchmarking sort_by_proximity,
as it's not exercised by the single-node perf-simple-query.

The benchmark shows performance improvement of over 20% (from about 71 ns to 56 ns
per call for 3 nodes vectors), mainly due to "calculate distance only once" which
pre-calculates the distance from the reference node for each replica once, rather than
each time to comparator is called by std::sort

* Improvement.  No backport needed

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13911/)
#### Build Details:
- Duration: 47 min
- Builder: i-003011695b0397d0b (m5d.12xlarge)

bhalevy commented on service/storage_proxy.cc:
```diff
@@ -6610,11 +6610,15 @@ void storage_proxy::sort_endpoints_by_proximity(const locator::effective_replica
+    if (topology.can_sort_by_proximity()) {
+        topology.do_sort_by_proximity(my_id, ids);
+    } else {
+        // FIXME: before dynamic snitch is implement put local address (if present) at the beginning
```
Fixed in v2 ([2489444](https://github.com/scylladb/scylladb/commit/2489444a9c86ed6622bbf8b11bda0b3e008d7e65))

bhalevy commented:
In v2 ([2489444](https://github.com/scylladb/scylladb/commit/2489444a9c86ed6622bbf8b11bda0b3e008d7e65)):
- rebased
- fixed grammar in comment as per https://github.com/scylladb/scylladb/pull/21958#discussion_r1889763924
- fixed build error of test/boost/network_topology_strategy_test by specializing the topology::distance template in locator/topology.cc
- locator/topology: sort_by_proximity: calculate distance only once
  - cleaned up patch to reduce diff to the minimum

tchaikov commented on test/manual/sort_by_proximity_test.cc:
```diff
@@ -0,0 +1,104 @@
+using clock_type = std::chrono::high_resolution_clock;
+
+// Called in a seastar thread
+void test_sort_by_proximity(clock_type::duration duration) {
```
probably we could leverage the facilities like `PERF_TEST_F()` provided by seastar? for instance, see https://github.com/scylladb/scylladb/blob/master/test/perf/perf_big_decimal.cc

bhalevy replied:
good idea, thanks

bhalevy replied:
In v3 ([bc3acd5](https://github.com/scylladb/scylladb/commit/bc3acd52e39c2f97e0a9ffc478e1a2c4db7f4051))

bhalevy commented:
In v3 ([bc3acd5](https://github.com/scylladb/scylladb/commit/bc3acd52e39c2f97e0a9ffc478e1a2c4db7f4051)):
- moved manual test to test/perf/perf_sort_by_proximity
- updated git log messages

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

#### Build Details:
- Duration: 14 min
- Builder: i-01e8caf90d8ad6b4a (m5ad.8xlarge)

avikivity commented on locator/topology.cc:
```diff
@@ -567,21 +567,27 @@ const endpoint_dc_rack& topology::get_location(const inet_address& ep) const {
     }
 }
 
 void topology::sort_by_proximity(locator::host_id address, host_id_vector_replica_set& addresses) const {
```
@gleb-cloudius do we still need duplicates of these functions?

gleb-cloudius replied:
Eventually it will be remove. You just queued a series that removes some of its users. May be last once. Need to check.

gleb-cloudius replied:
It looks like after the series to move streaming/repair to host id the function can be removed.

bhalevy replied:
There are still callers that use it with inet_address like: `range_streamer::get_all_ranges_with_sources_for`, `row_level_repair::sort_peer_nodes`, `storage_service::get_new_source_ranges`.

Cc @asias

gleb-cloudius replied:
@bhalevy did you check in the next branch?

bhalevy replied:
No.  Now I see 5a849b0a6a29ff9ebcd748a45b69bed73da19473 (Merge "Move more subsystems to use host ids instead of ips" from Gleb)

bhalevy replied:
I'll rebase on top of it once it's merged to master

avikivity commented on configure.py:
```diff
@@ -643,6 +643,7 @@ def find_ninja():
     'test/perf/perf_idl',
     'test/perf/perf_vint',
     'test/perf/perf_big_decimal',
```
> sort_by_proximity_topology.perf_sort_by_proximity      994665   996.540ns     0.900ns   995.639ns   998.420ns       0.000       0.000     13230.4      2982.9

13k instructions to sort by proximity?!

bhalevy replied:
I'll check if that's normalized by iteration, as each call does 64 iterations.

bhalevy replied:
It is normalized by the number of iterations in each run.
See https://github.com/scylladb/seastar/blob/master/tests/perf/perf_tests.cc#L332

avikivity replied:
So we need 1100 instructions to sort three items?

avikivity replied:
At least you get 4.5IPC.

btw std::sort is likely terrible for such low counts. Bubble sort would likely be better.

avikivity replied:
(or insertion sort)

avikivity commented on locator/topology.cc:
```diff
@@ -580,20 +580,31 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
 
 template <typename T>
 void topology::do_sort_by_proximity(T address, utils::small_vector<T, 3>& addresses) const {
```
2.8k cycles is much better than 13k cycles, but still outrageous.

What are we measuring here?

bhalevy replied:
sort_by_proximity of 15 nodes using a reference node that is one of them

avikivity commented on locator/topology.cc:
```diff
@@ -580,20 +580,31 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
+    }
+    std::ranges::sort(host_infos, [&](const info& i1, const info& i2) {
+        return i1.distance < i2.distance;
     });
```
std::ranges::sort(host_infos, std::ranges::less(), std::mem_fn(&host_info::distance));

avikivity commented on locator/topology.cc:
```diff
@@ -580,20 +580,31 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
+    for (const auto& id : addresses) {
+        const auto& loc1 = get_location(id);
+        host_infos.emplace_back(id, distance(address, loc, id, loc1));
+    }
```
Using std::range transformations and std::ranges::to would be faster since it can use the from_range_t constructor which avoids checks after every emplace_back (which the compiler may not be able to eliminate).

avikivity replied:
The distance calculations could be done during effective_replication_map preparation. In fact we could prepare all the shuffled variants and pick one randomly, for small replication factors. But out of scope for now.

bhalevy replied:
good idea, will do

bhalevy replied:
> The distance calculations could be done during effective_replication_map preparation. In fact we could prepare all the shuffled variants and pick one randomly, for small replication factors. But out of scope for now.

But the distance is from the coordinator node? Would we prepare the distance vectors against each possible node in the system?

avikivity commented on locator/topology.cc:
```diff
@@ -580,6 +581,9 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
 
 template <typename T>
 void topology::do_sort_by_proximity(T address, utils::small_vector<T, 3>& addresses) const {
+    static thread_local std::mt19937_64 random_engine(std::random_device{}());
```
Do we need some repeatable seed thing here for tests?

avikivity replied:
Can consider a simpler random engine like linear congruential (may or may not help).

mykaul replied:
There are others (xoshiro's) that are faster with smaller state size, I believe.

bhalevy replied:
I'm not sure it will make any difference, but we can have the random_engine live in topology
and expose a private method for the test to seed it.

avikivity commented on locator/topology.cc:
```diff
@@ -595,8 +599,15 @@ void topology::do_sort_by_proximity(T address, utils::small_vector<T, 3>& addres
+            }
+            shuffler = std::rotr(shuffler, 1);
+        }
+        *it++ = host_infos[i-1].id;
```
This is executed n-1 times, in the previous code n times.

bhalevy replied:
You're right, the last component must be copied too

avikivity commented on locator/topology.cc:
```diff
@@ -11,6 +11,7 @@
 #include <seastar/util/lazy.hh>
 #include <utility>
+#include <bit>
 
```
There's a huge discrepancy between the hit in instruction count and the hit in throughput.

bhalevy replied:
What metrics do you consider having a huge discrepancy?
```
Before:
test                                               iterations      median         mad         min         max      allocs       tasks        inst      cycles
sort_by_proximity_topology.perf_sort_by_proximity      994665   996.540ns     0.900ns   995.639ns   998.420ns       0.000       0.000     13230.4      2982.9

After:
sort_by_proximity_topology.perf_sort_by_proximity     4260690   236.851ns     0.523ns   236.327ns   241.161ns       1.000       0.000      2847.9       710.0
```

Instructions: 13230.4 / 2847.9 = 4.65
Cycles: 2982.9 / 710.0 = 4.20
Iterations: 4260690. / 994665 = 4.28

bhalevy replied:
Ah, I see what you mean.  The last patch causes a much bigger hit in cycles than instructions.

bhalevy commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -996,17 +1000,17 @@ SEASTAR_THREAD_TEST_CASE(test_topology_compare_endpoints) {
+        test_compare_endpoints(topo, address, a1, a2);
+        test_compare_endpoints(topo, address, a2, a1);
+
+        test_compare_endpoints(topo, bogus_address, bogus_address, bogus_address);
```
Yeah, can do

bhalevy replied:
https://github.com/scylladb/scylladb/pull/21976

bhalevy commented:
In v4 
- Rebased onto https://github.com/scylladb/scylladb/pull/21976 (and made draft in the meanwhile)
- used more std::ranges in so_sort_by_proximity
- moved random_engine to topology member and allow seeding from (perf) test
- locator/topology: do_sort_by_proximity: shuffle equal-distance replicas
  - moved copy back into in/out vector to std::ranges::copy
  - this improved cycles/iteration significantly, apparently due to better dcache utilization
- updated commit log messages with fresh perf numbers

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13962/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13962/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
#### Failed Tests (1/9):
- [test_topology_sort_by_proximity](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13962/testReport/junit/boost/network_topology_strategy_test/test_topology_sort_by_proximity) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_sort_by_proximity)
#### Build Details:
- Duration: 43 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
In v5 ([3ce6840](https://github.com/scylladb/scylladb/commit/3ce6840e73610212f8fc05a80d8f38c48409033a)):
- use 3 nodes for perf benchmark
- utils: small_vector: expose internal_capacity()
- locator/topology: do_sort_by_proximity: shuffle equal-distance replicas
  - fix bug found by unit test

bhalevy commented:
- rebased
- updated license
- ready to re-review now that https://github.com/scylladb/scylladb/pull/21976 is in

avikivity commented on test/perf/perf_sort_by_proximity.cc:
```diff
@@ -0,0 +1,73 @@
+struct sort_by_proximity_topology {
+    static constexpr size_t DCS = 1;
+    static constexpr size_t RACKS_PER_DC = 3;
+    static constexpr size_t NODES_PER_RACK = 1;
```
It would be more realistic to have a larger NODES_PER_RACK, so that we don't touch exactly the same nodes each time.

bhalevy replied:
Done in v6 ([d1490bb](https://github.com/scylladb/scylladb/commit/d1490bb7bfcdabe36ff50d88379e7edb7cbe2ddf))

avikivity commented on configure.py:
```diff
@@ -643,6 +643,7 @@ def find_ninja():
     'test/perf/perf_idl',
     'test/perf/perf_vint',
```
> number of cores:          32

I'm sure there are fewer.

Does the test run on all vcpus or just one?

bhalevy replied:
Tries also with --smp=1 and the number of iterations is the same, so it reports 32 but actually runs on 1

bhalevy replied:
Fixed in v6 ([d1490bb](https://github.com/scylladb/scylladb/commit/d1490bb7bfcdabe36ff50d88379e7edb7cbe2ddf))

avikivity commented on service/storage_proxy.cc:
```diff
@@ -6610,11 +6610,15 @@ void storage_proxy::sort_endpoints_by_proximity(const locator::effective_replica
+        if (it != ids.end() && it != ids.begin()) {
+            std::iter_swap(it, ids.begin());
+        }
     }
```
It would be better to enhance SimpleSnitch to contain rudimentary location metadata (like: I'm close and everyone else is far) to save this extra test. But not in scope here.

avikivity commented on locator/topology.cc:
```diff
@@ -573,16 +573,22 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
+    std::ranges::copy(addresses | std::views::transform([&] (locator::host_id id) {
+        const auto& loc1 = get_location(id);
+        return info{ id, distance(address, loc, id, loc1) };
+    }), std::back_inserter(host_infos));
```
std::back_inserter() is slow if the compiler can't figure out the reservation is sufficient. std::ranges::to is likely to be better, since it uses the from_ranges_t protocol to initialize the elements in-place.

bhalevy replied:
Used std::ranges::to in v6 ([d1490bb](https://github.com/scylladb/scylladb/commit/d1490bb7bfcdabe36ff50d88379e7edb7cbe2ddf))

avikivity commented on locator/topology.cc:
```diff
@@ -573,16 +573,22 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
+        const auto& loc1 = get_location(id);
+        return info{ id, distance(address, loc, id, loc1) };
+    }), std::back_inserter(host_infos));
+    std::ranges::sort(host_infos, std::ranges::less{}, std::mem_fn(&info::distance));
```
I see std::sort() switches to an insertion sort for small counts, so it's okay.

avikivity commented on locator/topology.cc:
```diff
@@ -585,6 +591,19 @@ void topology::do_sort_by_proximity(locator::host_id address, host_id_vector_rep
+            shuffler = std::rotr(shuffler, 1);
+        }
+        prev = it;
+    }
```
Idea: instead of a post-sort pass, add random bits to the distance metric. This will cause sort() to randomize the order of otherwise equidistant nodes.

bhalevy replied:
I'll try that.  I think that drawing 3 bits (since 64 is not divisible by 3) at a time out of the sampled 64-bit shuffle can do a good job.

bhalevy replied:
Tried with v6, and it performs slightly worse than inline shuffle in copy pass.
random-bits-concat:
```
test                                               iterations      median         mad         min         max      allocs       tasks        inst      cycles
    sort_by_proximity_topology.perf_sort_by_proximity    18556938    53.934ns     0.075ns    53.857ns    54.621ns       0.000       0.000       649.8       160.9
```

inline-shuffle:
```
sort_by_proximity_topology.perf_sort_by_proximity    19689561    50.195ns     0.119ns    50.076ns    51.145ns       0.000       0.000       622.5       150.6
```

avikivity commented on locator/topology.cc:
```diff
@@ -6,11 +6,14 @@
  */
 
+#include <bit>
+#include <ranges>
```
> locator/topology: do_sort_by_proximity: shuffle equal-distance replicas
> To improve balancing when reading in 1 < CL < ALL
> 
> This implementation has a moderate impact on
> the function performance in contrast to full
> std::shuffle of the vector before sorting it
> (especially with large number of nodes to sort).

That doesn't work anyway, since sort() may re-sort them back by address, if the implementation finds it convenient (it's not a stable sort).

bhalevy commented:
In v6 ([d1490bb](https://github.com/scylladb/scylladb/commit/d1490bb7bfcdabe36ff50d88379e7edb7cbe2ddf)):
- rebased (and resolved conflicts)
- test/perf: add perf_sort_by_proximity benchmark
  - allocate 3 nodes per rack
  - pre-calculate all possible replica sets with 1 replica per rack
  - measure sort of each replica set against each node
  - Note: there's no significant change in measured performance
- locator/topology: sort_by_proximity: calculate distance only once
  - use std::ranges::to

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14056/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14056/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2501/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/871/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/275/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13630/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5617/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14056/)
#### Build Details:
- Duration: 4 hr 12 min
- Builder: spider8.cloudius-systems.com
